File: individual-name-matching/java-examples/append-simkeys-to-file.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.HttpURLConnection;
import java.net.URLEncoder;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * Reads full names from a text file, calls Interzoid's Full Name Matching API
 * for each line, and writes the original value + SimKey to a CSV file.
 *
 * File name: append-simkeys-to-file-fullname.java
 * Run with:  java AppendFullnameSimkeysToFile
 */
class AppendFullnameSimkeysToFile {

    // Replace this with your API key from https://www.interzoid.com/manage-api-account
    private static final String API_KEY = "YOUR_API_KEY_HERE";

    // Hardcoded input and output file names (same as Go example)
    private static final String INPUT_FILE_NAME = "sample-input-file.txt";
    private static final String OUTPUT_FILE_NAME = "output.csv";

    // Simple POJO to hold parsed JSON response
    private static class ResponseData {
        String simKey;
        String code;
        String credits;
    }

    public static void main(String[] args) {
        try (
            BufferedReader reader = new BufferedReader(new FileReader(INPUT_FILE_NAME));
            PrintWriter writer = new PrintWriter(new BufferedWriter(new FileWriter(OUTPUT_FILE_NAME)))
        ) {
            String line;
            int lineNumber = 0;

            while ((line = reader.readLine()) != null) {
                lineNumber++;
                String originalValue = line;

                // Skip completely empty lines (same behavior as Go)
                if (originalValue.length() == 0) {
                    continue;
                }

                ResponseData result;
                try {
                    result = callApiForFullname(originalValue);
                } catch (Exception e) {
                    System.out.printf(
                        "API error on line %d (\"%s\"): %s%n",
                        lineNumber, originalValue, e.getMessage()
                    );
                    // On error, write original value and empty SimKey
                    writer.println(toCsvRow(originalValue, ""));
                    continue;
                }

                if (result == null) {
                    System.out.printf(
                        "No response for line %d (\"%s\")%n",
                        lineNumber, originalValue
                    );
                    writer.println(toCsvRow(originalValue, ""));
                    continue;
                }

                // (Optional) You could inspect result.code here if desired

                // Write original value and SimKey as a CSV row
                writer.println(toCsvRow(originalValue, result.simKey));
            }

            System.out.println("Done. Results written to " + OUTPUT_FILE_NAME);

        } catch (IOException e) {
            System.out.println("Error opening/creating files: " + e.getMessage());
        }
    }

    /**
     * Calls the Interzoid getfullnamematch API for a given full name.
     */
    private static ResponseData callApiForFullname(String fullName) throws IOException {
        String nameParam = URLEncoder.encode(fullName, StandardCharsets.UTF_8.toString());

        String apiUrl =
            "https://api.interzoid.com/getfullnamematch?license=" + API_KEY +
            "&fullname=" + nameParam;

        URL url = new URL(apiUrl);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");

        int status = conn.getResponseCode();
        BufferedReader in;
        if (status >= 200 && status < 300) {
            in = new BufferedReader(
                    new java.io.InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        } else {
            in = new BufferedReader(
                    new java.io.InputStreamReader(conn.getErrorStream(), StandardCharsets.UTF_8));
        }

        StringBuilder responseBuilder = new StringBuilder();
        String inputLine;
        while ((inputLine = in.readLine()) != null) {
            responseBuilder.append(inputLine);
        }
        in.close();
        conn.disconnect();

        String json = responseBuilder.toString().trim();
        if (json.isEmpty()) {
            return null;
        }

        ResponseData data = new ResponseData();
        data.simKey = extractJsonValue(json, "SimKey");
        data.code = extractJsonValue(json, "Code");
        data.credits = extractJsonValue(json, "Credits");
        return data;
    }

    /**
     * Very simple JSON value extractor for flat JSON like:
     * {"SimKey":"...","Code":"Success","Credits":"1"}
     *
     * This avoids external JSON libraries.
     */
    private static String extractJsonValue(String json, String key) {
        String quotedKey = "\"" + key + "\"";
        int keyIndex = json.indexOf(quotedKey);
        if (keyIndex == -1) {
            return "";
        }

        int colonIndex = json.indexOf(':', keyIndex + quotedKey.length());
        if (colonIndex == -1) {
            return "";
        }

        // Skip whitespace
        int valueStart = colonIndex + 1;
        while (valueStart < json.length() &&
               Character.isWhitespace(json.charAt(valueStart))) {
            valueStart++;
        }

        // Handle quoted string value
        if (valueStart < json.length() && json.charAt(valueStart) == '"') {
            valueStart++;
            StringBuilder sb = new StringBuilder();
            boolean escaping = false;

            for (int i = valueStart; i < json.length(); i++) {
                char c = json.charAt(i);
                if (escaping) {
                    sb.append(c);
                    escaping = false;
                } else if (c == '\\') {
                    escaping = true;
                } else if (c == '"') {
                    break;
                } else {
                    sb.append(c);
                }
            }
            return sb.toString();
        } else {
            // Fallback for non-string values (not expected here)
            int valueEnd = valueStart;
            while (valueEnd < json.length() &&
                   json.charAt(valueEnd) != ',' &&
                   json.charAt(valueEnd) != '}' &&
                   !Character.isWhitespace(json.charAt(valueEnd))) {
                valueEnd++;
            }
            return json.substring(valueStart, valueEnd).trim();
        }
    }

    /**
     * Escapes a single field for CSV:
     * - Wrap in quotes if it contains comma, quote, or newline
     * - Double any internal quotes
     */
    private static String csvEscape(String field) {
        if (field == null) {
            field = "";
        }
        boolean mustQuote = field.contains(",") ||
                            field.contains("\"") ||
                            field.contains("\n") ||
                            field.contains("\r");
        if (!mustQuote) {
            return field;
        }
        String escaped = field.replace("\"", "\"\"");
        return "\"" + escaped + "\"";
    }

    /**
     * Builds a single CSV row with two columns.
     */
    private static String toCsvRow(String col1, String col2) {
        return csvEscape(col1) + "," + csvEscape(col2);
    }
}

File: individual-name-matching/kotlin-examples/append-simkeys-to-file.kt
// append-simkeys-to-file.kt
//
// Reads names from a text file, one per line,
// calls the Interzoid Full Name Match API for each,
// and writes an output CSV containing the original name and the SimKey.
//
// Input:  sample-input-file.txt
// Output: output.csv
//
// Compile and run:
//   kotlinc append-simkeys-to-file.kt -include-runtime -d append-simkeys-to-file.jar
//   java -jar append-simkeys-to-file.jar

import java.io.File
import java.io.PrintWriter
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.io.BufferedReader
import java.io.InputStreamReader

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const val API_KEY = "YOUR_API_KEY_HERE"

// Hardcoded input and output file names
const val INPUT_FILE_NAME = "sample-input-file.txt"
const val OUTPUT_FILE_NAME = "output.csv"

fun main() {
    val inputFile = File(INPUT_FILE_NAME)
    if (!inputFile.exists()) {
        println("Error: input file \"$INPUT_FILE_NAME\" not found.")
        return
    }

    val outputFile = File(OUTPUT_FILE_NAME)

    PrintWriter(outputFile).use { writer ->
        var lineNumber = 0

        inputFile.useLines { lines ->
            lines.forEach { line ->
                lineNumber++
                val originalValue = line

                // Skip empty lines (optional)
                if (originalValue.isEmpty()) return@forEach

                // URL encode the name so it works safely
                val nameParam = URLEncoder.encode(originalValue, "UTF-8")

                // Build the API URL
                val apiURL =
                    "https://api.interzoid.com/getfullnamematch" +
                            "?license=$API_KEY" +
                            "&fullname=$nameParam"

                try {
                    // Make the GET request
                    val url = URL(apiURL)
                    val connection = url.openConnection() as HttpURLConnection
                    connection.requestMethod = "GET"

                    // Read body into a string
                    val responseBuilder = StringBuilder()
                    BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                        reader.forEachLine { responseLine ->
                            responseBuilder.append(responseLine)
                        }
                    }
                    val responseBody = responseBuilder.toString()

                    // Extract fields without any JSON library
                    val simKey = extractJsonValue(responseBody, "SimKey") ?: ""

                    // Write one row to CSV
                    writeCsvRow(writer, listOf(originalValue, simKey))

                } catch (e: Exception) {
                    println(
                        "API error on line $lineNumber " +
                                "(${quoteForLog(originalValue)}): ${e.message}"
                    )
                    // Write blank SimKey so row count remains consistent
                    writeCsvRow(writer, listOf(originalValue, ""))
                }
            }
        }
    }

    println("Done. Results written to $OUTPUT_FILE_NAME")
}

/**
 * Extracts a JSON field like "FieldName":"Value"
 * without external libraries.
 */
fun extractJsonValue(json: String, fieldName: String): String? {
    val regex = """"$fieldName"\s*:\s*"([^"]*)"""".toRegex()
    val match = regex.find(json)
    return match?.groups?.get(1)?.value
}

/**
 * Writes one CSV row, escaping commas/quotes safely.
 */
fun writeCsvRow(writer: PrintWriter, columns: List<String>) {
    val escapedColumns = columns.map { csvEscape(it) }
    writer.println(escapedColumns.joinToString(","))
}

/**
 * Escapes CSV safely by doubling quotes and wrapping with quotes.
 */
fun csvEscape(value: String): String {
    val escaped = value.replace("\"", "\"\"")
    return "\"$escaped\""
}

/**
 * Only used for log messages (not CSV).
 */
fun quoteForLog(value: String): String = "\"$value\""
